cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,85 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Public cuML interface
+### Motivation
+The public cuML interface is stateless for two main reasons:
+1. To ease the serialization of ML algorithm's state information (model, hyper-params, etc), enabling features such as easy pickling in the python layer.
+2. To provide proper C bindings for interfacing with languages that can't consume C++ APIs  directly.
+
+Thus, this section lays out guidelines for managing state along the API of cuML.
+
+### General guideline
+As mentioned before, functions exposed via the base cuML C++ layer must be stateless. Things that are OK to be exposed on the interface:
```

Comment:
I like the use of `base cuML C++ layer` here. Perhaps that wording should be used above.

Reply:
Yes, updated as per your previous suggestion.